### PR TITLE
First attempt at monitoring metrics using prometheus_client

### PR DESCRIPTION
**Pull request by AdolfoVillalobos**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## ⚠️ How does this PR impact the user? ⚠️
Describe (in plain English, not technical Jargon) how this improves the user experience. If you can't tie it back to a real tangible, user goal or describe it in plain english, it's a hint that this is likely not needed and is probably an "engineering nit". 

### GOOD:
"As a user, I need to serve models faster. This PR focuses on enabling speed gains by using GPUs"

### BAD:
"This PR enables GPUs". 
This is bad because the *user problem* is not clear... instead it just jumps to the solution without any context. 

PRs without this will not be merged.

---- 

## What does this PR do?

First attempt at solving #146 

## PR review

This implementation introduces basic monitoring metrics using prometheus_client to collect API metrics. The goal is to provide a foundation for further enhancements and discussions.

Implementing monitoring involves some decision-making and tradeoffs, so I start by:
1.  proposing changes. 
2. Highlight some limitations that I didn't want to solve right away before getting feedback
3. Summarize the key issues to discuss.

I hope this PR proves to be a useful ground for discussion :)

#### 1. Prometheus Integration:

1. Utilizes prometheus_client to collect metrics.
2. Integrates seamlessly with FastAPI, automatically creating a /metrics endpoint.
3. Compatible with multithreading environments.

#### 2. Implementation Approach:
1. The current approach involves some repetition. A more elegant solution would be to use decorators for methods like predict, encode, and decode.
2. Care must be taken with callback annotations and signatures to maintain API flexibility, especially in a multithreading context.
#### 3. Memory Management:

Given the multithreading environment, memory management for the Prometheus client may be necessary.
This is an area that requires further discussion to determine if it's something the team wants to address now.

#### 1. Metric Scope:
1. Only prediction elapsed time is measured currently.
1. Generalization to other metrics and custom metrics is straightforward but not yet implemented.
#### 2. Code Repetition:
1. The current implementation is somewhat repetitive.
1. Future work should focus on reducing this through the use of decorators.
#### 3. Memory Management:
1. Potential memory management issues with Prometheus in a multithreading environment need to be considered.

### Next Steps (Summary)
1. Discuss and decide on the best approach for extending metrics. Should we use prometheus?
1. Evaluate the use of decorators to reduce code repetition. This might involve some refactoring.
1. Consider the implications of memory management for Prometheus in our multithreading setup. Maybe a simpler solution would be best?

## Did you have fun?

Yess. I've been playing with this issue for a couple of hours and I think I have a better grasp of the project.

### CODE REVIEW

**Comment by aniketmaurya** [comment]
Hi @AdolfoVillalobos, great PR! I'm working on a callback system for LitServe which would act similar to middleware but gives you much more control over different parts of LitServe while serving a model.

I think this PR would be a great use case for the callback. I'm gonna create a PR for that soon and think that we can reorganize this PR to convert into a callback so that users can add or remove as required.

**Comment by AdolfoVillalobos** [comment]
That sounds really cool! I found myself thinking about the same problem while I was investigating the system. Good to know!

I will update this PR as soon as you submit yours.

I will be looking for other opportunities to contribute to the project in the meantime!

**Comment by AdolfoVillalobos** [comment]
@aniketmaurya I just realized the codecov/patch check on the CI is hanged. I can't cancel it. Should I worry? Cheers!

**Comment by aniketmaurya** [comment]
There is no need to worry about the hanging CI, @AdolfoVillalobos. It's probably some issue with codecov service.

**Comment by aniketmaurya** [comment]
hi @AdolfoVillalobos, callback system PR is merged. Let's do this!!

This PR can be refactored like the following: 

```python
from prometheus_client import Summary
from prometheus_client import multiprocess, CollectorRegistry, make_asgi_app

def make_metrics_app():
    generate_metrics_dir()
    registry = CollectorRegistry()
    multiprocess.MultiProcessCollector(registry)
    return make_asgi_app(registry=registry)

class PrometheusCallback(ls.Callback):
    def __init__(self):
        self.summary = Summary("PREDICT_METRIC", "A metric to summarize the predict execution times")

	def on_server_start(self, litserver):
        metrics_app = make_metrics_app()
		litserver.app.mount("/metrics", metrics_app)

    def on_before_predict(self, lit_api: "LitAPI"):
        t0 = time.perf_counter()
        self._start_time = t0

    def on_after_predict(self, lit_api: "LitAPI"):
        t1 = time.perf_counter()
        elapsed = t1 - self._start_time
        self.summary.observe(elapsed)
```

**Comment by williamFalcon** [comment]
ok, this actually needs to be broken up into 2 things @aniketmaurya , a prometheus logger and a self.log api

## A telemetry callback
The callback for telemetry (and i would also add decode, encode request, etc...).  

```python
class SpeedTelemetry(ls.Callback):

    def on_before_predict(self, lit_api: "LitAPI"):
        t0 = time.perf_counter()
        self._start_time = t0

    def on_after_predict(self, lit_api: "LitAPI"):
        t1 = time.perf_counter()
        elapsed = t1 - self._start_time
        self.log('predict time', elapsed)
```

## A logger
All the setup stuff is really at the logger level.    

```python
Server(callbacks=[SpeedTelemetry()], loggers=[Prometheus(), ...])
```

**Comment by AdolfoVillalobos** [comment]
> @williamFalcon created a [PR](https://github.com/Lightning-AI/LitServe/pull/284) to introduce the Logger API.

This is so cool. Thanks for the guidance. I will be taking a stab at this issue on the following days. Thanks!

**Comment by vrdn-23** [comment]
Thanks for the work adding compatibility for this @aniketmaurya !
I just wanted to understand the end goal of this PR if possible. Now that the functionality for callbacks and loggers have been landed,  would the end user be in charge of implementing and exposing their own metrics or does LitServer plan to have a set of metrics built-in with the option to extend with custom metrics if needed?
If it is the latter, are there a set of metrics that are determined to be made available or is that list still being discussed? 
cc @williamFalcon

**Comment by williamFalcon** [comment]
generally, metrics are very specific to each user/server/system. we will provide basic callbacks that people can add, but metrics can get expensive in terms of performance, so we want users to explicitly opt into the callbacks they use. 

but adding a metric you care about is now super straightforward with self.log(…) inside the litapi.

**Comment by vrdn-23** [comment]
@williamFalcon sounds great! Thanks for the clarification! I guess in that case is there any estimated timeline for when the next release of LitServe is going to be cut?

**Comment by vrdn-23** [comment]
I am currently trying to implement some custom metrics for our team and we would be interested in a metric that is similar to `queue_size` which provides the current number of requests waiting in the queue. But from what I can tell the `request_queue` property isn't exposed via the LitAPI interface, so is there some other way that I'm missing that can help calculate this?

**Comment by aniketmaurya** [comment]
@vrdn-23 although you can use [queue.qsize()](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue.qsize) method, it is approximate and not reliable. 

Also, a cleaner way to this would be to have a [middleware](https://lightning.ai/docs/litserve/features/callbacks-middleware#using-middleware-in-litserve) with a counter that increment the counter when it enters the request and decrement it after processing the request.

**Comment by aniketmaurya** [comment]
You can use prometheus with LitServe now. Please follow the docs [here](https://lightning.ai/docs/litserve/features/logger#prometheus).
